Add `rating`, `volatility`, and contest history

Ninjaclasher opened the pull request:
Adds `rating`, `volatility`, and contest history to the user API.

Xyene commented on judge/views/api/api_v1.py:
```diff
@@ -109,18 +109,34 @@ def api_v1_user_list(request):
+    for participation in (ContestParticipation.objects.filter(user=profile, virtual=0, contest__is_public=True, contest__is_private=False)
+                                  .order_by('-contest__end_time')):
+        contest_history[participation.contest.key] = {
+            'rating': participation.rating.rating if hasattr(participation, 'rating') else None,
```
This and the line above will execute a query for every contest the user has participated in (= potentially many queries). You probably want to use something like `values_list` here instead.

Xyene commented on judge/views/api/api_v1.py:
```diff
@@ -109,18 +109,34 @@ def api_v1_user_list(request):
         'rank': rank
     } for username, name, points, rank in queryset})
 
-
```
Functions in global scope should leave 2 newlines between themselves.

Xyene commented on judge/views/api/api_v1.py:
```diff
@@ -136,3 +152,4 @@ def api_v1_user_submissions(request, user):
         'status': sub['status'],
         'result': sub['result'],
     } for sub in subs.values('id', 'problem__code', 'time', 'memory', 'points', 'language__key', 'status', 'result')})
+
```
Extra newline.

Xyene commented on judge/views/api/api_v1.py:
```diff
@@ -5,7 +5,7 @@
 
 from dmoj import settings
-from judge.models import Contest, Problem, Profile, Submission, ContestTag
+from judge.models import Contest, ContestParticipation, Problem, Profile, Submission, ContestTag
```
Sort imports.

quantum5 commented on judge/views/api/api_v1.py:
```diff
@@ -114,13 +114,30 @@ def api_v1_user_info(request, user):
-    })
+    }
+
+    last_rating = profile.ratings.order_by('-contest__end_time').first()
```
Can you just reuse the query you are using below? Otherwise, at least get rid of the sort on that query.

Ninjaclasher commented:
Updated, although in my opinion this looks less clean. Another possibility would be to remove the `order_by` sort and use `.last()` to get the last rating, but I'm not sure if it will handle well with rerating contests (`rating` `id`s will be changed).

Ninjaclasher commented:
That's an interesting feature that I was not aware of. However, I don't think that would work in this case, since the  last `rating` or `volatility` might be `None` (if the contest was not rated). I've also removed the sort query, and reverted how `last_rating` is retrieved.
